libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
-      connectionsTimeout: connectionsTimeout)
-
-  return transport
-
-method start*(
-  self: TcpTransport,
-  addrs: seq[MultiAddress]) {.async.} =
-  ## listen on the transport
-  ##
-
-  if self.running:
-    warn "TCP transport already running"
-    return
-
-  await procCall Transport(self).start(addrs)
-  trace "Starting TCP transport"
-  trackCounter(TcpTransportTrackerName)
+    T: typedesc[TcpTransport],
+    flags: set[ServerFlags] = {},
+    upgrade: Upgrade,
+    connectionsTimeout = 10.minutes,
+): T {.public.} =
+  T(
+    flags: flags,
+    clientFlags:
+      if ServerFlags.TcpNoDelay in flags:
+        {SocketFlags.TcpNoDelay}
+      else:
+        default(set[SocketFlags])
+    ,
+    upgrader: upgrade,
+    networkReachability: NetworkReachability.Unknown,
+    connectionsTimeout: connectionsTimeout,
+  )
+
+method start*(self: TcpTransport, addrs: seq[MultiAddress]): Future[void] =
+  ## Start transport listening to the given addresses - for dial-only transports,
+  ## start with an empty list
+
+  # TODO remove `impl` indirection throughout when `raises` is added to base
+
+  proc impl(
+      self: TcpTransport, addrs: seq[MultiAddress]
+  ): Future[void] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if self.running:
+      warn "TCP transport already running"
+      return
 
-  for i, ma in addrs:
-    if not self.handles(ma):
-      trace "Invalid address detected, skipping!", address = ma
-      continue
+    trace "Starting TCP transport"
 
     self.flags.incl(ServerFlags.ReusePort)
-    let server = createStreamServer(
-      ma = ma,
-      flags = self.flags,
-      udata = self)
 
-    # always get the resolved address in case we're bound to 0.0.0.0:0
-    self.addrs[i] = MultiAddress.init(
-      server.sock.getLocalAddress()
-    ).tryGet()
+    var supported: seq[MultiAddress]
+    var initialized = false
+    try:
+      for i, ma in addrs:
+        if not self.handles(ma):
+          trace "Invalid address detected, skipping!", address = ma
+          continue
 
-    self.servers &= server
+        let
+          ta = initTAddress(ma).expect("valid address per handles check above")
+          server =
+            try:
+              createStreamServer(ta, flags = self.flags)
+            except common.TransportError as exc:
+              raise (ref TcpTransportError)(msg: exc.msg, parent: exc)
+
+        self.servers &= server
+
+        trace "Listening on", address = ma
+        supported.add(
+          MultiAddress.init(server.sock.getLocalAddress()).expect(
+            "Can init from local address"
+          )
+        )
+
+      initialized = true
+    finally:
+      if not initialized:
+        # Clean up partial success on exception
+        await noCancel allFutures(self.servers.mapIt(it.closeWait()))
+        reset(self.servers)
 
-    trace "Listening on", address = ma
+    try:
+      await procCall Transport(self).start(supported)
```

Comment:
You're probably trying to set "running" only after all the code above ran without issues, but it seems strange to call a parent method only at the end. It will be confusing to see the start log only after the start happened and the associated logs.

Reply:
this is a weakness of the inheritance / method structure of the code - ie the semantics of `running` are such that we should not mark the transport as "running" until it has been fully started (all transports are running), but the way transport in general is implemented is that calling the base method is the way to "set" running. 

The previous way of setting `running` to true before actually starting the listeners is  part of the buggy behavior that led to race conditions.

refactoring this flow would require refactoring _all_ transport implementations which is beyond the scope of this PR.

Reply:
If nothing important is being done in the parent and we don't plan to change it, perhaps we can stop calling it and set the flag manually.

Reply:
that requires refactoring all transports (ws/tor too) - something beyond the scope here

Reply:
why refactoring? it's stop calling the parent and setting the flags in the child.

Reply:
because to do it properly, one would remove the code from the base call (or really, change the structure so that the base contains functionality for this that doesn't involve inheritance) - then the other transports need to be updated - notably though, the base method needs refactoring anyway for `raises`, so it can be done together then.

it would be confusing to have some implementations call the base and some not